pkg/solana/logpoller/log_poller.go

```diff
@@ -0,0 +1,316 @@
+package logpoller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+	"slices"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+var (
+	ErrFilterNameConflict = errors.New("filter with such name already exists")
+)
+
+type ORM interface {
+	InsertFilter(ctx context.Context, filter Filter) (id int64, err error)
+	SelectFilters(ctx context.Context) ([]Filter, error)
+	DeleteFilters(ctx context.Context, filters []Filter) error
+	MarkFilterDeleted(ctx context.Context, id int64) (err error)
+}
+
+type LogPoller struct {
+	services.StateMachine
+	lggr logger.SugaredLogger
+	orm  ORM
+
+	filtersByName     map[string]Filter
+	filtersByAddress  map[PublicKey]map[EventSignature][]Filter
+	filtersToBackfill []Filter
+	filtersToDelete   []Filter // populated on start from db and pruned on first iteration of run
+	filtersMutex      sync.RWMutex
+	loadedFilters     atomic.Bool
+
+	chStop services.StopChan
+	wg     sync.WaitGroup
+}
+
+func NewLogPoller(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+	return &LogPoller{
+		orm:  orm,
+		lggr: logger.Sugared(logger.Named(lggr, "LogPoller")),
+
+		filtersByName:    make(map[string]Filter),
+		filtersByAddress: map[PublicKey]map[EventSignature][]Filter{},
+	}
+}
+
+func (lp *LogPoller) Start(context.Context) error {
+	return lp.StartOnce("LogPoller", func() error {
+		lp.wg.Add(2)
+		go lp.run()
+		go lp.backgroundWorkerRun()
+		return nil
+	})
+}
+
+func (lp *LogPoller) Close() error {
+	return lp.StopOnce("LogPoller", func() error {
+		close(lp.chStop)
+		lp.wg.Wait()
+		return nil
+	})
+}
+
+func (lp *LogPoller) run() {
+	defer lp.wg.Done()
+	ctx, cancel := lp.chStop.NewCtx()
+	defer cancel()
+
+	var blocks chan struct {
+		BlockNumber int64
+		Logs        any // to be defined
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case block := <-blocks:
+			lp.filtersMutex.Lock()
+			filtersToBackfill := lp.filtersToBackfill
+			lp.filtersToBackfill = nil
+			lp.filtersMutex.Unlock()
+			// TODO: NONEVM-916 parse, filters and persist logs
+			// NOTE: removal of filters occurs in the separate goroutine, so there is a chance that upon insert
+			// of log corresponding filter won't be present in the db. Ensure to refilter and retry on insert error
+			for _, filter := range filtersToBackfill {
+				go lp.startFilterBackfill(ctx, filter, block.BlockNumber)
+			}
+		}
+	}
+}
+
+func (lp *LogPoller) backgroundWorkerRun() {
+	defer lp.wg.Done()
+	ctx, cancel := lp.chStop.NewCtx()
+	defer cancel()
+
+	pruneFilters := services.NewTicker(time.Minute + 618*time.Millisecond) // try to minimize collisions with one-second period
```

Comment:
`NewTicker` adds a +-10% random jitter offset to whatever value is passed, so adding an extra offset is probably not necessary.

I guess the 1s period you're referring is the timer we'll use for the main logpoller loop?